### PR TITLE
Solution by DK

### PR DESCRIPTION
**Pull request by DimaK123**
[DEMO LINK](https://Dimak123.github.io/react_todo-app-add-and-delete/)

### CODE REVIEW

File: src/components/Filter/FilterParams.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export enum FilterParams {
+  all = 'all',
+  completed = 'completed',
+  active = 'active',
```
[the remark this answers is not in the document]
Comment (DimaK123) [reply src/components/Filter/FilterParams.ts]: All done, but I can't figure out how to make server requests not constant, but only when changes are made on the page?

File: src/App.tsx [hunk 2]
```diff
@@ -1,24 +1,203 @@
...
+    state = true,
+  ) => {
+    setError(state);
+    setErrorFunc(state);
+
+    if (state === false) {
+      return;
+    }
+
+    setTimeout(() => handleErrorState(setErrorFunc, !state), 3000);
```
Comment (Anastasiia-Svintsova) [inline src/App.tsx:48]: You should clear all timers on the component unmount 
https://bobbyhadz.com/blog/react-cleartimeout

File: src/App.tsx [hunk 3]
```diff
@@ -1,24 +1,203 @@
...
+  const [isClearAllCompleted, setClearAllCompleted] = useState(false);
+
+  const handleErrorState = (
+    setErrorFunc: (state: boolean) => void,
+    state = true,
+  ) => {
+    setError(state);
+    setErrorFunc(state);
+
+    if (state === false) {
```
Comment (Anastasiia-Svintsova) [inline src/App.tsx:44]: You can use logical operators instead
```suggestion
    if (!state) {
```

File: src/App.tsx [hunk 4]
```diff
@@ -1,24 +1,203 @@
...
+  const [isInputLocked, setLockInput] = useState(false);
+  const [tempTodo, setTempTodo] = useState<string | null>(null);
+  const [isClearAllCompleted, setClearAllCompleted] = useState(false);
+
+  const handleErrorState = (
+    setErrorFunc: (state: boolean) => void,
+    state = true,
+  ) => {
+    setError(state);
+    setErrorFunc(state);
```
Comment (Anastasiia-Svintsova) [inline src/App.tsx:42]: Why do you need 2 identical states? Could you refactor it and leave only 1?

File: src/App.tsx [hunk 5]
```diff
@@ -1,24 +1,203 @@
...
+    }
+
+    setTimeout(() => handleErrorState(setErrorFunc, !state), 3000);
+  };
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(result => setTodos(result))
+      .catch(() => {
+        handleErrorState(setGetError);
```
Comment (Anastasiia-Svintsova) [inline src/App.tsx:55]: Why do you need to pass the function here? Just set the error inside the catch

File: src/App.tsx [hunk 6]
```diff
@@ -1,24 +1,203 @@
...
+      });
+  }, [todos]);
+
+  useEffect(() => {
+    getActiveTodos(USER_ID).then(result => setActiveTodos(result.length));
+
+    getCompletedTodos(USER_ID).then(result => setCompletedTodos(result.length));
+  }, [filterParam, todos]);
```
Comment (Anastasiia-Svintsova) [inline src/App.tsx:63]: Not the best way to get values. Use useMemo() instead

File: src/App.tsx [hunk 7]
```diff
@@ -1,24 +1,203 @@
...
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const visibleTodos = todos.filter(todo => {
```
Comment (Anastasiia-Svintsova) [inline src/App.tsx:69]: Use useMemo() here

File: src/components/TempTodo/TempTodo.tsx [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+
+type Props = {
+  title: string
+};
+
+export const TempTodo:React.FC<Props> = ({ title }) => {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (Anastasiia-Svintsova) [inline src/components/TempTodo/TempTodo.tsx:8]: Don't disable eslint

**Review by Anastasiia-Svintsova — CHANGES_REQUESTED** [review]
- The user shouldn't be able to add a todo with an empty title. Fix it, please 

https://user-images.githubusercontent.com/99403198/235292666-3932c3f9-f742-43e0-9049-dc40147bf951.mov

- Look, here is infinite fetching. You should fetch it at the beginning and then only after changes. It really decreases the productivity 

https://user-images.githubusercontent.com/99403198/235292740-99f5fa8c-7220-4dad-b1b4-4a54f13f0e75.mov

**Comment by DimaK123** [comment]
> * The user shouldn't be able to add a todo with an empty title. Fix it, please
> 
> 
> Screen.Recording.2023-04-29.at.11.21.10.mov
> 
>     * Look, here is infinite fetching. You should fetch it at the beginning and then only after changes. It really decreases the productivity
> 
> 
> Screen.Recording.2023-04-29.at.11.23.02.mov

So, I've fixed

File: src/App.tsx (hunk 2 above)
Comment (DimaK123) [reply src/App.tsx:48]: I don't know how to do it using useEffect, so I implemented the timer deletion in a function.

File: src/App.tsx (hunk 3 above)
Comment (DimaK123) [reply src/App.tsx:44]: Done

File: src/App.tsx (hunk 4 above)
Comment (DimaK123) [reply src/App.tsx:42]: what do you mean?
I use only one state param, and pass it to both functions.

File: src/App.tsx (hunk 5 above)
Comment (DimaK123) [reply src/App.tsx:55]: i need pass correct error set-function to the handleErrorState function to show message 'Error, can\'t get todos from server', so i use this function

File: src/App.tsx (hunk 6 above)
Comment (DimaK123) [reply src/App.tsx:63]: Done

File: src/App.tsx (hunk 7 above)
Comment (DimaK123) [reply src/App.tsx:69]: i can't do it cause the basic value for the visibleTodos it empty array, so if i try to use useMemo here, i get error from typescript

File: src/components/TempTodo/TempTodo.tsx (hunk 8 above)
Comment (DimaK123) [reply src/components/TempTodo/TempTodo.tsx:8]: Done

**Comment by DimaK123** [comment]
> * The user shouldn't be able to add a todo with an empty title. Fix it, please
> 
> 
> Screen.Recording.2023-04-29.at.11.21.10.mov
> 
>     * Look, here is infinite fetching. You should fetch it at the beginning and then only after changes. It really decreases the productivity
> 
> 
> Screen.Recording.2023-04-29.at.11.23.02.mov

fixed

File: src/App.tsx (hunk 4 above)
Comment (OleziO) [reply src/App.tsx:42]: You don't need two states for error, just do it with one state.

File: src/api/todos.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    title,
+    completed: false,
+  });
+};
+
+export const deleteTodos = (id: number) => {
+  return client.delete(`/todos/${id}`);
+};
+
+// Add more methods here
```
Comment (OleziO) [inline src/api/todos.ts:28]: Remove all comments from the code.

File: src/components/Todo/Todo.tsx [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+}) => {
+  const { completed, title, id } = todoItem;
+
+  const [isLoading, setLoading] = useState(false);
+
+  const handleDeleteTodos = (todoId: typeof id) => {
+    setLoading(true);
+
+    deleteTodos(todoId)
+      .then(() => setLoading(false))
```
[the remark this answers is not in the document]
Comment (OleziO) [reply src/components/Todo/Todo.tsx:33]: Not resolved.

**Review by OleziO — CHANGES_REQUESTED** [review]
Delete does not work.

File: src/api/todos.ts (hunk 9 above)
Comment (DimaK123) [reply src/api/todos.ts]: Done

**Review by polosanya — CHANGES_REQUESTED** [review]
It is still not possible to remove a todo.
Items left count also doesn't work correctly
todos

**Comment by DimaK123** [comment]
> It is still not possible to remove a todo. Items left count also doesn't work correctly todos [ todos ](https://user-images.githubusercontent.com/96409563/235422391-19a4225b-992b-4928-9d9c-cf07ced8c188.gif) [ ](https://user-images.githubusercontent.com/96409563/235422391-19a4225b-992b-4928-9d9c-cf07ced8c188.gif)

it's seems i fogot to run 'npm run deploy'

**Pull request closed without merging**